pyronear/datasets/wildfire/fire_labeler.py

```diff
@@ -0,0 +1,150 @@
+import re
+from itertools import combinations
+
+import numpy as np
+
+
+class FireLabeler:
+    """Automatically labelize WildFire dataset based on video descriptions
+
+    It will create a new column containing Fire ids that try to identify the videos
+    illustrating same real fire.
+
+    An instance of the Labeler is bound to a dataframe but can be
+    run several times, in order to vary the window size for instance.
+
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        This DataFrame should:
+            - be indexed from 0 to (max number of videos-1) (range-like)
+            - contain the video descriptions in the uploading order.
+                The closer the upload, the more likely videos represent same fire
+            - have a column named 'description' containing the description of the videos
+
+    window_size: in (default=10)
+        Count of video descriptions to use to determine if they represent same fire.
+
+    Attributes
+    ----------
+    n_videos_total: int
+        Count of videos (rows found in dataframe)
+
+    n_windows: int
+        Count of windows
+
+    _n_singletons: int >= 0
+        Count of videos not grouped with at least one other
+        Learnt attribute, available after run()
+
+    Examples
+    --------
+    df = pd.read_csv("WildFire.csv", index_col=0)
+    fire_labeler = FireLabeler(df, window_size=30)
+    fire_labeler.run()
+    df_on_fire = fire_labeler.get_new_dataframe(column_name='fire_id')
+
+    df_on_fire_short = (fire_labeler.reset(window_size=6)
+                                    .run()
+                                    .get_dataframe(column_name='fire_id'))
+    """
+    def __init__(self, df, window_size=10):
+        self.df = df
+        self.window_size = window_size
+
+        self.reset()
+
+    def reset(self, window_size=None):
+        """Reset the labeler
+        Reset fire ids previously found(if any) and set again all dimensions
+        Especially useful, if we want to try another window size"""
+        self.window_size = window_size or self.window_size
+
+        self.n_videos_total = self.df.shape[0]
+        self.n_windows = self.n_videos_total - self.window_size + 1  # n_windows + windows_size < n_videos_total
+
+        # Store new column for fire ids starting at 0 (-1 for unassigned)
+        self.fire_ids = np.full((self.n_videos_total), -1)
+        return self
+
+    def run(self):
+        """ Run the labelisation of the fire depending on the descriptions
+
+        For every combination of descriptions(strings) in every sliding window
+        guess if they belong to same fire (fire_id)
+
+        Note: algo complexity can be improved (ex: by memoizing sliding combinations).
+        But, for now, processing is fast enough(<1min) when used with large window-size<100
```

Comment:
maybe I missed the logic but why a window size is needed and why it uses windows of fixed size for consecutive entries? Couldn't it compare pairs of descriptions and keep only the different "groups" ?

Reply:
Thanks,
You are absolutely right. It _tries_ to group according to same fire so we minimize leakage *in time* between datasets splits(train/val/test). Surely, same camera lookouts may be spread among each split; to a certain extent, the trained model may not generalized very well, but practically it should be fined-tuned or retrained with data collected from the field. This should thwart generalization gap.

Reply:
Thanks for your question.Concerning the window size:
- Yes it is needed, because we are looking for pairs not only in consecutive entries (it would be `windows_size` of 2) but also entries quite far away. You can browse the tests to see how it can benefit when dealing with long dependency. For instance Matching `KingFire` and `King Fire` requires to have a pair containing these two, but this is not happening consecutively.
- Comparing _every pair_ of descriptions would be too costly: complexity grows to the square of the count of videos you take into account:
![image](https://user-images.githubusercontent.com/6331909/70142976-be32db00-169a-11ea-8857-f6654a0c8df4.png)
So for a window of `n = 100 videos`, it roughly makes  *5000 iterations*,
but for a window of `n = 1000 videos` (all), it will require *500 000 iterations*, so 100 times more. In this case it will require more optimization, and I think this is an overkill for our problem right now, because we are trying to group videos representing same fire,videos that were uploaded, a priori, in the same order they were recorded.

Does this help?

Reply:
Yeah, I thought that indeed you tried to make use of the sequence. But I did not mean to compare every pair in practice. The idea was something inspired in merge sort: you compare two descriptions and if they match, you keep one as a representative of the group, otherwise you keep both. 

It can end up as n**2 if you compare all pairs and there are no similar ones but you can also compare only the adjacent ones to reduce the size. 

Just an idea. if it works already, then fine.

Reply:
Thanks for the suggestion. You're right, when designing the algorithm, I wanted to be able to use the proximity of the descriptions as a reinforcement factor to detect matching. In the end, I did not use this because I think this is good enough. Your suggestion is also interesting but it would be more difficult to make use of the sequence. For now and for this dataset, I don't see what it could practically improve.